src/DurableTask.AzureStorage/OrchestrationMemoryManager.cs

```
@@ -0,0 +1,195 @@
+        }
+
+        async Task CurrentMemoryMonitor()
+        {
+            this.settings.Logger.OrchestrationMemoryManagerInfo(
+                this.storageAccountName,
+                this.taskHub,
+                $"Starting background monitor of currently allocated memory.");
+
+            while (this.isStarted == 1)
+            {
+                this.currentlyAllocatedMemory = Process.GetCurrentProcess().PrivateMemorySize64;
+
+                await Task.Delay(10, this.currentMemoryCancellationTokenSource.Token);
+            }
+
+            this.settings.Logger.OrchestrationMemoryManagerInfo(
+                this.storageAccountName,
+                this.taskHub,
+                $"Background monitor for currenly allocated memory completed.");
+        }
+
+        public async Task<OrchestrationHistory> GetHistoryEventsAsync(ITrackingStore trackingStore, string instanceId, string executionId, CancellationToken cancellationToken)
+        {
+            bool memoryReserved = false;
+            long memorySize = 0;
+
+            try
+            {
+                if (this.settings.MemoryThrottleSettings.UseOrchestrationHistoryLoadThrottle)
+                {
+                    OrchestrationState state = await trackingStore.GetStateAsync(
+                            instanceId,
+                            executionId,
+                            false);
+
+                    memoryReserved = await this.WaitUntilMemoryReservedAsync(state, cancellationToken);
+                    memorySize = state.Size;
+                }
+
+                if (memoryReserved || !this.settings.MemoryThrottleSettings.UseOrchestrationHistoryLoadThrottle)
+                {
+                    OrchestrationHistory history = await trackingStore.GetHistoryEventsAsync(
+                    instanceId,
+                    executionId,
+                    cancellationToken);
+
+                    return history;
+                }
+
+                return null;
```

When do we expect to return `null` here?
If it's expected in a valid scenario: can we please add a comment documenting it.
If it's not supposed to happen, should we just throw an exception?